iocore/net/QUICNetVConnection.cc

```diff
@@ -575,13 +616,43 @@ QUICNetVConnection::state_connection_established(int event, Event *data)
   return EVENT_CONT;
 }
 
+void
+QUICNetVConnection::_prepare_to_close_or_drain()
+{
+  this->_unschedule_packet_write_ready();
+  this->_unschedule_closing_timeout();
+  this->next_inactivity_timeout_at = 0;
+  this->next_activity_timeout_at   = 0;
+  this->inactivity_timeout_in      = 0;
+  this->active_timeout_in          = 0;
+
+  this->_loss_detector->handleEvent(QUIC_EVENT_LD_SHUTDOWN, nullptr);
+}
+
 int
 QUICNetVConnection::state_connection_closing(int event, Event *data)
 {
   SCOPED_MUTEX_LOCK(lock, this->mutex, this_ethread());
 
   QUICErrorUPtr error = QUICErrorUPtr(new QUICNoError());
   switch (event) {
+  case QUIC_EVENT_SHUTDOWN:
+    this->_prepare_to_close_or_drain();
+    // Shutdown loss detector
+    // In the closing state, only a packet containing a closing frame can be sent.
+    this->_stream_manager->close_all_streams(VC_EVENT_ERROR);
+
+    // TODO: find rto from cc
+    // rto = this->_cc->get_rto()
+    // 7.9.1. Closing and Draining Connection States
+    // The closing and draining connection states exist to ensure that connections close
+    // cleanly and that delayed or reordered packets are properly discarded. These states SHOULD
+    // persist for three times the current Retransmission Timeout (RTO) interval as defined in [QUIC-RECOVERY].
+    this->nh->add_to_active_queue(this);
+    set_active_timeout(3 * HRTIME_MSECONDS(200));
+    this->_state_closing_send_packet();
```

Comment:
Why do we need to call `_state_closing_send_packet()` here? It's a kind of event handler. I mean it is supposed to be called from event handler when the handler receives a `QUIC_EVENT_PACKET_WRITE_READY`.

Reply:
`_state_closing_send_packet ` means send close frame immediately.  spec says we should send close frame when we enter the close state(we sent it in _switch_to_closing_state in the original code).

Reply:
These paragraphs, right?

> An endpoint sends a closing frame, either CONNECTION_CLOSE or
   APPLICATION_CLOSE, to terminate the connection immediately.  Either
   closing frame causes all streams to immediately become closed; open
   streams can be assumed to be implicitly reset.

   > After sending a closing frame, endpoints immediately enter the
   closing state.

I read them differently. My understanding is that a closing frame will be sent if an endpoint want to close the connection immediately.

`_switch_to_closing_state` ensures the last sentence, because this is the only place we send closing frames.

Reply:
Yes, your understanding is correct.

Reply:
So what we need to do immediately is not sending a frame but state change. In original code, calling `_switch_to_closing_state` will enqueue the frame and switch the connection state immediately. The frame will be sent when we handle next WRITE_READY. If we call `_state_closing_send_packet` here, the connection may not be ready to write.